src/prioritizedTaskExecutor.ts

```diff
@@ -29,20 +34,51 @@ export class PrioritizedTaskExecutor {
    * @param priority The priority of the task
    * @param fn The function that accepts the callback, which must be called upon the task completion.
    */
-  execute(priority: number, fn: Function) {
+  async execute(priority: number, fn: Function) {
+    await this.lock.acquire()
     if (this.currentPoolSize < this.maxPoolSize) {
       this.currentPoolSize++
-      fn(() => {
+      fn(async () => {
+        await this.lock.acquire()
```

Comment:
This looks quite dangerous. If this callback gets executed and `await` right away you'd get into a deadlock, won't you?

That being said, I don't know much about this part of MPT. I was just casually looking at this because of your message on discord.

Reply:
I have to update it because the tests are failing, but I don't think this will deadlock: if it's locked it means that some other Promise is running, which will unlock it at some point.

Reply:
Never mind you are right 😅